torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,408 @@
+# type: ignore
+
+from __future__ import annotations
+
+import contextlib
+from collections import defaultdict
+
+import torch
+from torch.utils.data import Dataset
+
+from torchvision import datasets
+from torchvision.prototype import datapoints
+from torchvision.prototype.transforms import functional as F
+
+__all__ = ["wrap_dataset_for_transforms_v2"]
+
+
+# TODO: naming!
+def wrap_dataset_for_transforms_v2(dataset):
+    dataset_cls = type(dataset)
+    wrapper_factory = WRAPPER_FACTORIES.get(dataset_cls)
+    if wrapper_factory is None:
+        # TODO: If we have documentation on how to do that, put a link in the error message.
+        msg = f"No wrapper exist for dataset class {dataset_cls.__name__}. Please wrap the output yourself."
+        if dataset_cls in datasets.__dict__.values():
+            msg = (
+                f"{msg} If an automated wrapper for this dataset would be useful for you, "
+                f"please open an issue at https://github.com/pytorch/vision/issues."
+            )
+        raise ValueError(msg)
+    return VisionDatasetDatapointWrapper(dataset, wrapper_factory(dataset))
+
+
+class WrapperFactories(dict):
+    def register(self, dataset_cls):
+        def decorator(wrapper_factory):
+            self[dataset_cls] = wrapper_factory
+            return wrapper_factory
+
+        return decorator
+
+
+WRAPPER_FACTORIES = WrapperFactories()
```

Comment:
perhaps add a comment for those like myself who will forget and wonder why it's done this way:

> We need factories of wrappers and not just wrappers, because some of the wrappers are **instance-dependent** (as they use instance attributes as defined by users). The factories thus need to generate the proper wrapper at execution time.